ENT-717 Show generic info message on enterprise course enrollment page

zubair-arbi opened the pull request:
@asadiqbal08 @saleem-latif 

**Description:** Show generic info message on enterprise course enrollment page in case there was an error getting course or course run information.

**JIRA:** [ENT-717](https://openedx.atlassian.net/browse/ENT-717)

**Dependencies:** N/A

**Merge deadline:** 3 November, 2017 

**Installation instructions:** Install `edx-enterprise` from this branch `zub/ENT-717-course-enrollment-show-generic-messages-for-error` in `edx-platform`.


**Testing instructions:**

1. Create a SAML IDP for while enabling the setting "Drop existing session"
2. Create enterprise with the above IDP
3. Create a related catalog in model `EnterpriseCustomerCatalog`
4. Access the enterprise course enrollment url with the catalog query parameter (value should be the UUID of above created `EnterpriseCustomerCatalog` records) with url like:
http://zubair.localhost:8000/enterprise/72416e52-8c77-4860-9584-15e5b06220fb/course/course-v1:edX+DemoX+Demo_Course/enroll/?catalog=6eca3efb-f3a0-4c08-806f-c6e6b65d61cb
and verify that page renders without any error.

5. Now try to access same page with invalid course id:
http://zubair.localhost:8000/enterprise/72416e52-8c77-4860-9584-15e5b06220fb/course/course-v1:edX+DemoX+Invalid_Course/enroll/?catalog=6eca3efb-f3a0-4c08-806f-c6e6b65d61cb

6. Verify that now learner sees a generic info message `"Something happened.
This course is not available. Please start over and select a different course."` without any course information on the page.
<img width="841" alt="screen shot 2017-10-30 at 12 41 07 pm" src="https://user-images.githubusercontent.com/5072991/32159901-2b71c544-bd71-11e7-8977-59bb6cfe7f85.png">


**Merge checklist:**

- [ ] Check that the versions of the requirements in the `platform-master.in` file match edx-platform.
- [ ] New requirements are in the right place (`base.in` if only used in enterprise; in the correct `platform-****.in` files if they're hosted in edx-platform)
- [ ] Regenerate requirements with `make upgrade && make requirements` (and make sure to fix any errors).
  **DO NOT** just add dependencies to `requirements/*.txt` files.
- [ ] Called `make static` for webpack bundling if any static content was updated.
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [ ] Documentation updated (not only docstrings)
- [x] Commits are (reasonably) squashed
- [ ] Translations are updated
- [ ] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is finished.
- [ ] Delete working branch (if not needed anymore)
- [ ] edx-platform PR (be sure to include edx-platform requirements upgrades that were present in this PR)

**Author concerns:** List any concerns about this PR - inelegant 
solutions, hacks, quick-and-dirty implementations, concerns about 
migrations, etc.

douglashall commented on enterprise/templates/enterprise/enterprise_course_enrollment_page.html:
```diff
@@ -21,67 +21,69 @@
-        {% if course_enrollable %}
-          <h2 class="course-confirmation-title">{{ confirmation_text }}</h2>
-        {% endif %}
+        {% if course_and_course_run_available %}
```
We are beginning to add a lot of conditional logic to this template. This seems like a good opportunity to clean up both the view and the template by rendering rendering a separate template for the error conditions.

asadiqbal08 commented on enterprise/views.py:
```diff
@@ -633,53 +637,80 @@ def get_enterprise_course_enrollment_page(
+            if course['owners']:
+                # The owners key contains the organizations associated with the course.
+                # We pick the first one in the list here to meet UX requirements.
+                organization = course['owners'][0]
```
nit: [:1]

zubair-arbi replied:
This statement is protected by the condition `if course['owners']` so no `IndexError` will be raised and also we need an individual item for organization not a list.

zubair-arbi commented:
@douglashall Yes, I have started working on it today. I think it would be great if we can handle refactoring in a separate story.

douglashall commented:
Awesome. I don't know why we need another story. I think adjusting the points on the existing story is fine.

zubair-arbi commented:
@douglashall I have added a separate template for enterprise course enrollment error page and updated the view accordingly. Please let me know if this is what we were thinking.

douglashall commented on enterprise/templates/enterprise/enterprise_course_enrollment_error_page.html:
```diff
@@ -0,0 +1,24 @@
+  </div>
+  {% with index=0 %}
+    {% course_modal %}
+  {% endwith %}
```
Should this be here?

zubair-arbi replied:
Nope :)

zubair-arbi commented on enterprise/templates/enterprise/enterprise_course_enrollment_error_page.html:
```diff
@@ -0,0 +1,24 @@
+        <p class="partnered-text">{{ welcome_text }}</p>
+        <p class="partnered-text">
+          {% autoescape off %}{{ enterprise_welcome_text }}{% endautoescape %}
+        </p>
```
Done

douglashall commented on enterprise/templates/enterprise/enterprise_course_enrollment_error_page.html:
```diff
@@ -0,0 +1,18 @@
+  <div class="enterprise-container">
+    <div class="row">
+      <div class="col-3">
+        {% include "enterprise_course_enrollment_sidebar_column.html" %}
```
Can you name this partial `_enterprise_customer_sidebar.html`?

zubair-arbi replied:
Sure.

zubair-arbi replied:
Done.

zubair-arbi commented:
Squashed and rebased onto master from commit d344e7c1ace7f553248778241411d60586ae66ce.

douglashall commented on enterprise/templates/enterprise/_enterprise_customer_sidebar.html:
```diff
@@ -0,0 +1,7 @@
+</p>
+<p class="partnered-text">
+    {% autoescape off %}{{ enterprise_welcome_text }}{% endautoescape %}
+</p>
```
Can you also include this partial in [grant_data_sharing_permissions.html](https://github.com/edx/edx-enterprise/blob/master/enterprise/templates/enterprise/grant_data_sharing_permissions.html#L27)?

zubair-arbi replied:
Done.

douglashall commented on enterprise/templates/enterprise/_enterprise_customer_sidebar.html:
```diff
@@ -0,0 +1,7 @@
+<img class="enterprise-logo" src="{{ enterprise_customer.branding_configuration.logo.url }}" alt="{{ enterprise_customer.name }}"/>
```
Should we check for the existence of a branding_configuration the way we are doing [here](https://github.com/edx/edx-enterprise/blob/master/enterprise/templates/enterprise/grant_data_sharing_permissions.html#L27)?

zubair-arbi replied:
Yes, Its better to not display the enterprise logo placeholder instead of the actual logo if there is no branding available for the enterprise. Thanks for pointing this out.

zubair-arbi commented:
@douglashall I have addressed your feedback.
@saleem-latif @asadiqbal08 can some one of you please review the changes?

douglashall commented:
@zubair-arbi Looking good! Let's get this one merged.
